File: api/language_detector.php
<?php
/**
 * Language Detection for Chatbot
 * Detects English, Kinyarwanda, and French
 */

// Kinyarwanda common words
$KINYARWANDA_WORDS = [
    'mwaramutse', 'mwiriwe', 'muraho', 'habari', 'amakuru', 'iki', 'nde', 'ute',
    'inyumba', 'umuntu', 'igitabo', 'amafaranga', 'ubwoko', 'ikinini', 'icyuma',
    'murakoze', 'urakoze', 'asante', 'ahubwo', 'yego', 'oya', 'nta', 'niba',
    'ariko', 'kandi', 'cyangwa', 'kubera', 'nkuko', 'uko', 'aho', 'hano',
    'hari', 'nari', 'niba', 'nkuko', 'ubwoko', 'icyiciro', 'inzira', 'umwanya',
    'umunsi', 'igice', 'igipimo', 'igihe', 'igikorwa', 'igitabo', 'igikoni',
    'igikoresho', 'igitambara', 'igitero', 'igitonda', 'igituba', 'igitugu',
    'kugura', 'kuguza', 'kugurisha', 'kugurwa', 'kunywa', 'kurya', 'kuryamwo',
    'gusoma', 'gusomeka', 'kwiyandikisha', 'kwiyandikira', 'gukora', 'gukoresha',
    'gusobanura', 'gusaba', 'gusabye', 'gusabyeho',
];

// French common words
$FRENCH_WORDS = [
    'bonjour', 'bonsoir', 'bonne', 'nuit', 'salut', 'coucou', 'allo', 'allô',
    'merci', 'merci beaucoup', 'merci bien', 'de rien', 'au revoir', 'adieu',
    'oui', 'non', 'peut-être', 'peut', 'peux', 'pouvez', 'pouvez-vous',
    'je', 'tu', 'il', 'elle', 'nous', 'vous', 'ils', 'elles',
    'mon', 'ma', 'mes', 'ton', 'ta', 'tes', 'son', 'sa', 'ses',
    'notre', 'nos', 'votre', 'vos', 'leur', 'leurs',
    'un', 'une', 'des', 'le', 'la', 'les', 'l',
    'et', 'ou', 'mais', 'donc', 'car', 'parce', 'que', 'qui', 'quoi',
    'où', 'quand', 'comment', 'pourquoi', 'combien', 'quel', 'quelle',
    'quels', 'quelles', 'lequel', 'laquelle', 'lesquels', 'lesquelles',
    'ce', 'cet', 'cette', 'ces', 'celui', 'celle', 'ceux', 'celles',
    'ça', 'cela', 'ceci', 'là', 'ici', 'là-bas', 'là-haut', 'là-dedans',
    'veux', 'vouloir', 'voulez', 'voudriez', 'aimerais', 'aimer', 'aimez',
    'besoin', 'cherche', 'chercher', 'cherchez', 'trouve', 'trouver', 'trouvez',
    'prix', 'coût', 'coûte', 'coûter', 'coûtez', 'livraison', 'livrer', 'livrez',
    'commande', 'commander', 'commandez', 'paiement', 'payer', 'payez',
    'retour', 'retourner', 'retournez', 'produit', 'produits', 'article', 'articles',
    'magasin', 'boutique', 'store', 'shop', 'commerce', 'client', 'clients',
    'vendeur', 'vendeurs', 'seller', 'sellers', 'marchand', 'marchands',
    'combien', 'combien ça coûte', 'quel est le prix', 'quel prix',
    'frais', 'frais de port', 'frais de livraison', 'gratuit', 'gratuite',
    'gratuits', 'gratuites', 'free', 'sans frais', 'jour', 'jours',
    'jour ouvrable', 'jours ouvrables', 'semaine', 'semaines', 'mois', 'année',
    'rapide', 'rapides', 'lent', 'lents', 'cher', 'chère', 'chers', 'chères',
    'bon marché', 'bon', 'bonne', 'bons', 'bonnes', 'cheap', 'qualité',
    'qualités', 'quality', 'mauvais', 'mauvaise', 'excellent', 'excellente',
    'excellents', 'excellentes', 'terrible', 'terribles', 'bien', 'mal',
    'très', 'beaucoup', 'peu', 'assez', 'trop', 'pas', 'ne', 'rien',
    'jamais', 'toujours', 'encore', 'déjà', 'bientôt', 'maintenant',
    'aujourd', 'aujourd hui', 'aujourd\'hui', 'hier', 'demain',
];

// English common words
$ENGLISH_WORDS = [
    'hello', 'hi', 'hey', 'good', 'morning', 'afternoon', 'evening', 'night',
    'thank', 'thanks', 'thank you', 'thanks a lot', 'thanks so much',
    'yes', 'no', 'maybe', 'can', 'could', 'would', 'should', 'will', 'shall',
    'i', 'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'her', 'us', 'them',
    'my', 'your', 'his', 'her', 'its', 'our', 'their', 'mine', 'yours',
    'a', 'an', 'the', 'and', 'or', 'but', 'so', 'because', 'that', 'which', 'who',
    'what', 'where', 'when', 'why', 'how', 'how much', 'how many',
    'this', 'that', 'these', 'those', 'here', 'there',
    'want', 'need', 'like', 'love', 'hate', 'prefer', 'choose', 'select', 'pick',
    'buy', 'purchase', 'sell', 'shop', 'shopping', 'store',
    'product', 'products', 'item', 'items', 'thing', 'things', 'stuff',
    'price', 'cost', 'money', 'pay', 'payment', 'charge', 'fee', 'fees',
    'delivery', 'shipping', 'ship', 'deliver', 'send', 'sent', 'arrive', 'arrived',
    'order', 'orders', 'track', 'tracking', 'return', 'returns',
    'refund', 'refunds', 'cancel', 'cancellation', 'help', 'support',
    'question', 'questions', 'answer', 'answers', 'information', 'info',
    'know', 'understand', 'see', 'look', 'find', 'search', 'show', 'display', 'list', 'browse',
    'fast', 'quick', 'slow', 'expensive', 'cheap', 'good', 'bad',
    'excellent', 'terrible', 'very', 'much', 'little', 'enough', 'too',
    'not', 'nothing', 'never', 'always', 'still', 'already', 'soon', 'now',
    'today', 'yesterday', 'tomorrow',
];

function detect_language($text) {
    global $KINYARWANDA_WORDS, $FRENCH_WORDS, $ENGLISH_WORDS;
    
    if (empty(trim($text))) {
        return 'english'; // default
    }
    
    $text_lower = strtolower(trim($text));
    preg_match_all('/\b\w+\b/', $text_lower, $matches);
    $words = $matches[0];
    
    if (empty($words)) {
        return 'english';
    }
    
    // Count matches for each language
    $kinyarwanda_count = 0;
    $french_count = 0;
    $english_count = 0;
    
    foreach ($words as $word) {
        if (in_array($word, $KINYARWANDA_WORDS)) {
            $kinyarwanda_count++;
        }
        if (in_array($word, $FRENCH_WORDS)) {
            $french_count++;
        }
        if (in_array($word, $ENGLISH_WORDS)) {
            $english_count++;
        }
    }
    
    // Determine language by highest count
    if ($kinyarwanda_count > $french_count && $kinyarwanda_count > $english_count) {
        return 'kinyarwanda';
    } elseif ($french_count > $english_count) {
        return 'french';
    } else {
        return 'english';
    }
}

// Multilingual response templates
$MULTILINGUAL_RESPONSES = [
    'greeting' => [
        'english' => [
            "Hello! Welcome to AI-Powered Chatbot For E-commerce Support. How can I help you today?",
            "Hi there! I'm your AI shopping assistant. What can I do for you?",
            "Hey! Great to see you. How can I assist you today?"
        ],
        'kinyarwanda' => [
            "Mwaramutse! Murakaza neza kuri AI-Powered Chatbot For E-commerce Support. Ndi iki nkwifuza?",
            "Mwiriwe! Ndi umukozi wacu wa shopping. Ndi iki nkwifuza?",
            "Muraho! Neza kuguha. Ndi iki nkwifuza?"
        ],
        'french' => [
            "Bonjour! Bienvenue sur le Chatbot IA pour le support e-commerce. Comment puis-je vous aider?",
            "Salut! Je suis votre assistant d'achat IA. Que puis-je faire pour vous?",
            "Hé! Ravi de vous voir. Comment puis-je vous aider?"
        ]
    ],
    'thanks' => [
        'english' => [
            "You're welcome! Happy to help.",
            "Glad I could assist! Is there anything else you need?",
            "My pleasure! Let me know if you need anything else."
        ],
        'kinyarwanda' => [
            "Ubwenge! Nari neza kubagenzi.",
            "Nari neza kubagenzi! Hari indi nkwifuza?",
            "Nari neza! Mbabarire niba hari indi nkwifuza."
        ],
        'french' => [
            "De rien! Heureux de pouvoir aider.",
            "Heureux d'avoir pu vous aider! Y a-t-il autre chose dont vous avez besoin?",
            "Mon plaisir! Faites-moi savoir si vous avez besoin d'autre chose."
        ]
    ],
    'product_search' => [
        'english' => [
            "I can help you find products! What category are you looking for?",
            "Sure! We have 1,161+ products. What are you looking for?",
            "Let me search our catalog for you. What product do you need?"
        ],
        'kinyarwanda' => [
            "Nshobora kubagenzi gushaka ibicuruzwa! Ubwoko biki bushaka?",
            "Yego! Dufite ibicuruzwa 1,161+. Ndi iki bushaka?",
            "Nzashaka mu katalogo yacu. Ndi iki bushaka?"
        ],
        'french' => [
            "Je peux vous aider à trouver des produits! Quelle catégorie cherchez-vous?",
            "Bien sûr! Nous avons plus de 1 161 produits. Que cherchez-vous?",
            "Laissez-moi chercher dans notre catalogue. Quel produit avez-vous besoin?"
        ]
    ],
    'product_price' => [
        'english' => [
            "I can check the price for you! Which product are you asking about?",
            "Our prices range from RWF 1,000 to RWF 5,000,000. Which product?",
            "Let me find the price for that product right away!"
        ],
        'kinyarwanda' => [
            "Nshobora kubagenzi gushaka igiciro! Iki bicuruzwa?",
            "Igiciro cyacu gihera kuri RWF 1,000 kugeza RWF 5,000,000. Iki bicuruzwa?",
            "Nzashaka igiciro cy'icyo bicuruzwa!"
        ],
        'french' => [
            "Je peux vérifier le prix pour vous! Quel produit demandez-vous?",
            "Nos prix vont de RWF 1 000 à RWF 5 000 000. Quel produit?",
            "Laissez-moi trouver le prix de ce produit tout de suite!"
        ]
    ],
    'order_track' => [
        'english' => [
            "I can help you track your order! Please provide your order number.",
            "To track your order, please share your order number.",
            "Let me check your order status. What's your order number?"
        ],
        'kinyarwanda' => [
            "Nshobora kubagenzi gushaka agaciro k'agaciro. Mbabarire numero y'agaciro.",
            "Gushaka agaciro, mbabarire numero y'agaciro.",
            "Nzashaka agaciro. Numero y'agaciro ari iki?"
        ],
        'french' => [
            "Je peux vous aider à suivre votre commande! Veuillez fournir votre numéro de commande.",
            "Pour suivre votre commande, veuillez partager votre numéro de commande.",
            "Laissez-moi vérifier l'état de votre commande. Quel est votre numéro de commande?"
        ]
    ],
    'delivery_time' => [
        'english' => [
            "Delivery to Kigali takes 1-2 business days. Other provinces take 2-4 days.",
            "We deliver within 1-2 days in Kigali and 2-4 days in other provinces.",
            "Standard delivery: Kigali 1-2 days, Provinces 2-4 days, Remote areas up to 7 days."
        ],
        'kinyarwanda' => [
            "Kugerageza i Kigali gutwara iminsi 1-2. Ibindi bigo gutwara iminsi 2-4.",
            "Tugerageza mu minsi 1-2 i Kigali no mu minsi 2-4 mu bigo.",
            "Kugerageza ibisanzwe: Kigali iminsi 1-2, Ibindi bigo iminsi 2-4, Ahantu hahuje kugeza iminsi 7."
        ],
        'french' => [
            "La livraison à Kigali prend 1-2 jours ouvrables. Les autres provinces prennent 2-4 jours.",
            "Nous livrons dans 1-2 jours à Kigali et 2-4 jours dans les autres provinces.",
            "Livraison standard: Kigali 1-2 jours, Provinces 2-4 jours, Zones éloignées jusqu'à 7 jours."
        ]
    ],
    'shipping_fee' => [
        'english' => [
            "Free shipping on all orders! No minimum required.",
            "All orders get FREE shipping, no minimum purchase!",
            "Shipping is free for all orders."
        ],
        'kinyarwanda' => [
            "Kugerageza kubuntu kuri buri gurwa!",
            "Buri gurwa rifite kugerageza kubuntu.",
            "Kugerageza kubuntu kuri buri gurwa."
        ],
        'french' => [
            "Livraison gratuite sur toutes les commandes!",
            "Toutes les commandes bénéficient d'une livraison GRATUITE.",
            "La livraison est gratuite pour toutes les commandes."
        ]
    ],
    'payment_methods' => [
        'english' => [
            "We accept: Cash on Delivery, MTN MoMo, Airtel Money, Visa/Mastercard, Bank Transfer.",
            "Payment options: COD, MTN Mobile Money, Airtel Money, Card, Bank Transfer.",
            "You can pay via Cash on Delivery, Mobile Money (MTN/Airtel), Card, or Bank Transfer."
        ],
        'kinyarwanda' => [
            "Dukwemera: Amafaranga mu gihe cy'ugerageza, MTN MoMo, Airtel Money, Visa/Mastercard, Kwishyura mu banki.",
            "Ubwoko bw'amafaranga: COD, MTN Mobile Money, Airtel Money, Karita, Kwishyura mu banki.",
            "Urashobora kwishyura mu gihe cy'ugerageza, Mobile Money (MTN/Airtel), Karita, cyangwa Kwishyura mu banki."
        ],
        'french' => [
            "Nous acceptons: Paiement à la livraison, MTN MoMo, Airtel Money, Visa/Mastercard, Virement bancaire.",
            "Options de paiement: COD, MTN Mobile Money, Airtel Money, Carte, Virement bancaire.",
            "Vous pouvez payer par Paiement à la livraison, Mobile Money (MTN/Airtel), Carte ou Virement bancaire."
        ]
    ],
    'return_policy' => [
        'english' => [
            "You can return items within 7 days of delivery for a full refund.",
            "Return policy: 7 days from delivery date. Items must be unused and in original packaging.",
            "We offer hassle-free returns within 7 days. Contact support for details."
        ],
        'kinyarwanda' => [
            "Urashobora kugarura ibicuruzwa mu minsi 7 nyuma y'ugerageza.",
            "Politiki y'ugarura: iminsi 7kuva ku minsi y'ugerageza. Ibicuruzwa byombi bitakoresha no mu nzira y'ibanze.",
            "Dufite ugarura burinshi mu minsi 7. Menya inyandiko z'inyubako."
        ],
        'french' => [
            "Vous pouvez retourner les articles dans les 7 jours suivant la livraison pour un remboursement complet.",
            "Politique de retour: 7 jours à partir de la date de livraison. Les articles doivent être inutilisés et dans l'emballage d'origine.",
            "Nous offrons des retours sans tracas dans les 7 jours. Contactez le support pour plus de détails."
        ]
    ]
];

function get_multilingual_response($intent, $language = 'english') {
    global $MULTILINGUAL_RESPONSES;
    
    if (isset($MULTILINGUAL_RESPONSES[$intent][$language])) {
        $responses = $MULTILINGUAL_RESPONSES[$intent][$language];
        return $responses[array_rand($responses)];
    }
    
    // Fallback to English
    if (isset($MULTILINGUAL_RESPONSES[$intent]['english'])) {
        $responses = $MULTILINGUAL_RESPONSES[$intent]['english'];
        return $responses[array_rand($responses)];
    }
    
    return "How can I help you?";
}

File: chatbot-ml/language_detector.py
"""
Language Detection Module for Chatbot
Detects English, Kinyarwanda, and French
"""

import re

# Common Kinyarwanda words and patterns
KINYARWANDA_WORDS = {
    'mwaramutse', 'mwiriwe', 'muraho', 'habari', 'amakuru', 'iki', 'nde', 'ute',
    'inyumba', 'inyamaswa', 'inyama', 'inyanya',
    'umuntu', 'umunsi', 'umwanya', 'umwari',
    'abantu', 'abakazi', 'abagabo', 'abana', 'abakungu', 'abakobwa',
    'igitabo', 'igice', 'igipimo', 'igihe', 'igikorwa', 'igikoni',
    'igikoresho', 'igitambara', 'igitero', 'igitondo', 'igituba', 'igitugu',
    'igitukuza', 'igitukuzwa', 'igituza', 'igituzwa', 'igitwa', 'igitwe', 'igitweho',
    'ibikinini', 'ibikoresho', 'ibigitabo', 'ibigitero', 'ibigitonda', 'ibigituba',
    'ibigitugu', 'ibigitukuza', 'ibigitukuzwa', 'ibigituza', 'ibigituzwa', 'ibigitwa',
    'ibigitwe', 'ibigitweho',
    'amafaranga', 'ubwoko', 'ikinini', 'icyuma', 'icyumba',
    'murakoze', 'urakoze', 'asante', 'ahubwo', 'yego', 'oya', 'nta', 'niba',
    'ariko', 'kandi', 'cyangwa', 'kubera', 'nkuko', 'uko', 'aho', 'hano',
    'hari', 'nari', 'icyiciro', 'inzira', 'inziranzira',
    'kugura', 'kuguza', 'kugurisha', 'kugurwa',
    'kunywa', 'kurya', 'kuryamwo',
    'gusoma', 'gusomeka',
    'kwiyandikisha', 'kwiyandikira',
}

# Common French words
FRENCH_WORDS = {
    'bonjour', 'bonsoir', 'bonne', 'nuit', 'salut', 'coucou', 'allo', 'allô',
    'merci', 'merci beaucoup', 'merci bien', 'de rien', 'au revoir', 'adieu',
    'oui', 'non', 'peut-être', 'peut', 'peux', 'pouvez', 'pouvez-vous',
    'je', 'tu', 'il', 'elle', 'nous', 'vous', 'ils', 'elles',
    'mon', 'ma', 'mes', 'ton', 'ta', 'tes', 'son', 'sa', 'ses',
    'notre', 'nos', 'votre', 'vos', 'leur', 'leurs',
    'un', 'une', 'des', 'le', 'la', 'les', 'l',
    'et', 'ou', 'mais', 'donc', 'car', 'parce', 'que', 'qui', 'quoi',
    'où', 'quand', 'comment', 'pourquoi', 'combien', 'quel', 'quelle',
    'quels', 'quelles', 'lequel', 'laquelle', 'lesquels', 'lesquelles',
    'ce', 'cet', 'cette', 'ces', 'celui', 'celle', 'ceux', 'celles',
    'ça', 'cela', 'ceci', 'là', 'ici', 'là-bas', 'là-haut', 'là-dedans',
    'veux', 'vouloir', 'voulez', 'voudriez', 'aimerais', 'aimer', 'aimez',
    'besoin', 'besoin', 'besoin', 'besoin', 'besoin', 'besoin', 'besoin',
    'cherche', 'chercher', 'cherchez', 'cherchais', 'cherchait', 'cherchaient',
    'trouve', 'trouver', 'trouvez', 'trouvais', 'trouvait', 'trouvaient',
    'prix', 'coût', 'coûte', 'coûter', 'coûtez', 'coûtais', 'coûtait',
    'livraison', 'livrer', 'livrez', 'livrais', 'livrait', 'livraient',
    'commande', 'commander', 'commandez', 'commandais', 'commandait', 'commandaient',
    'paiement', 'payer', 'payez', 'payais', 'payait', 'payaient',
    'retour', 'retourner', 'retournez', 'retournais', 'retournait', 'retournaient',
    'produit', 'produits', 'article', 'articles', 'item', 'items',
    'magasin', 'boutique', 'store', 'shop', 'commerce', 'commerces',
    'client', 'clients', 'customer', 'customers', 'acheteur', 'acheteurs',
    'vendeur', 'vendeurs', 'seller', 'sellers', 'marchand', 'marchands',
    'combien', 'combien ça coûte', 'quel est le prix', 'quel prix',
    'frais', 'frais de port', 'frais de livraison', 'frais de shipping',
    'gratuit', 'gratuite', 'gratuits', 'gratuites', 'free', 'sans frais',
    'jour', 'jours', 'jour ouvrable', 'jours ouvrables', 'jour de travail',
    'semaine', 'semaines', 'mois', 'mois', 'année', 'années',
    'rapide', 'rapides', 'rapide', 'rapide', 'rapide', 'rapide', 'rapide',
    'lent', 'lents', 'lent', 'lent', 'lent', 'lent', 'lent', 'lent',
    'cher', 'chère', 'chers', 'chères', 'expensive', 'expensive', 'expensive',
    'bon marché', 'bon', 'bonne', 'bons', 'bonnes', 'cheap', 'cheap', 'cheap',
    'qualité', 'qualités', 'quality', 'qualities', 'bon', 'bonne', 'bons',
    'mauvais', 'mauvaise', 'mauvais', 'mauvaise', 'mauvais', 'mauvaise', 'mauvais',
    'excellent', 'excellente', 'excellents', 'excellentes', 'excellent', 'excellent',
    'terrible', 'terribles', 'terrible', 'terrible', 'terrible', 'terrible', 'terrible',
    'bien', 'bien', 'bien', 'bien', 'bien', 'bien', 'bien', 'bien', 'bien',
    'mal', 'mal', 'mal', 'mal', 'mal', 'mal', 'mal', 'mal', 'mal', 'mal',
    'très', 'très', 'très', 'très', 'très', 'très', 'très', 'très', 'très',
    'beaucoup', 'beaucoup', 'beaucoup', 'beaucoup', 'beaucoup', 'beaucoup', 'beaucoup',
    'peu', 'peu', 'peu', 'peu', 'peu', 'peu', 'peu', 'peu', 'peu', 'peu',
    'assez', 'assez', 'assez', 'assez', 'assez', 'assez', 'assez', 'assez', 'assez',
    'trop', 'trop', 'trop', 'trop', 'trop', 'trop', 'trop', 'trop', 'trop', 'trop',
    'pas', 'pas', 'pas', 'pas', 'pas', 'pas', 'pas', 'pas', 'pas', 'pas', 'pas',
    'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne', 'ne',
    'rien', 'rien', 'rien', 'rien', 'rien', 'rien', 'rien', 'rien', 'rien', 'rien',
    'jamais', 'jamais', 'jamais', 'jamais', 'jamais', 'jamais', 'jamais', 'jamais',
    'toujours', 'toujours', 'toujours', 'toujours', 'toujours', 'toujours', 'toujours',
    'encore', 'encore', 'encore', 'encore', 'encore', 'encore', 'encore', 'encore',
    'déjà', 'déjà', 'déjà', 'déjà', 'déjà', 'déjà', 'déjà', 'déjà', 'déjà',
    'bientôt', 'bientôt', 'bientôt', 'bientôt', 'bientôt', 'bientôt', 'bientôt',
    'maintenant', 'maintenant', 'maintenant', 'maintenant', 'maintenant', 'maintenant',
    'aujourd', 'aujourd hui', 'aujourd\'hui', 'aujourd hui', 'aujourd hui',
    'hier', 'hier', 'hier', 'hier', 'hier', 'hier', 'hier', 'hier', 'hier',
    'demain', 'demain', 'demain', 'demain', 'demain', 'demain', 'demain', 'demain',
}

# Common English words (for comparison)
ENGLISH_WORDS = {
    'hello', 'hi', 'hey', 'good', 'morning', 'afternoon', 'evening', 'night',
    'thank', 'thanks', 'thank you', 'thanks a lot', 'thanks so much',
    'yes', 'no', 'maybe', 'can', 'could', 'would', 'should', 'will', 'shall',
    'i', 'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'her', 'us', 'them',
    'my', 'your', 'his', 'her', 'its', 'our', 'their', 'mine', 'yours', 'his',
    'a', 'an', 'the', 'and', 'or', 'but', 'so', 'because', 'that', 'which', 'who',
    'what', 'where', 'when', 'why', 'how', 'how much', 'how many', 'which', 'what',
    'this', 'that', 'these', 'those', 'here', 'there', 'there', 'where', 'where',
    'want', 'need', 'like', 'love', 'hate', 'prefer', 'choose', 'select', 'pick',
    'buy', 'purchase', 'sell', 'shop', 'shopping', 'store', 'shop', 'shop', 'shop',
    'product', 'products', 'item', 'items', 'thing', 'things', 'stuff', 'stuff',
    'price', 'cost', 'money', 'pay', 'payment', 'charge', 'fee', 'fees', 'fees',
    'delivery', 'shipping', 'ship', 'deliver', 'send', 'sent', 'arrive', 'arrived',
    'order', 'orders', 'order', 'order', 'order', 'order', 'order', 'order', 'order',
    'track', 'tracking', 'track', 'track', 'track', 'track', 'track', 'track', 'track',
    'return', 'returns', 'return', 'return', 'return', 'return', 'return', 'return',
    'refund', 'refunds', 'refund', 'refund', 'refund', 'refund', 'refund', 'refund',
    'cancel', 'cancellation', 'cancel', 'cancel', 'cancel', 'cancel', 'cancel', 'cancel',
    'help', 'help', 'help', 'help', 'help', 'help', 'help', 'help', 'help', 'help',
    'support', 'support', 'support', 'support', 'support', 'support', 'support', 'support',
    'question', 'questions', 'question', 'question', 'question', 'question', 'question',
    'answer', 'answers', 'answer', 'answer', 'answer', 'answer', 'answer', 'answer',
    'information', 'info', 'information', 'information', 'information', 'information',
    'know', 'know', 'know', 'know', 'know', 'know', 'know', 'know', 'know', 'know',
    'understand', 'understand', 'understand', 'understand', 'understand', 'understand',
    'see', 'see', 'see', 'see', 'see', 'see', 'see', 'see', 'see', 'see', 'see',
    'look', 'look', 'look', 'look', 'look', 'look', 'look', 'look', 'look', 'look',
    'find', 'find', 'find', 'find', 'find', 'find', 'find', 'find', 'find', 'find',
    'search', 'search', 'search', 'search', 'search', 'search', 'search', 'search',
    'show', 'show', 'show', 'show', 'show', 'show', 'show', 'show', 'show', 'show',
    'display', 'display', 'display', 'display', 'display', 'display', 'display', 'display',
    'list', 'list', 'list', 'list', 'list', 'list', 'list', 'list', 'list', 'list',
    'browse', 'browse', 'browse', 'browse', 'browse', 'browse', 'browse', 'browse',
    'fast', 'quick', 'slow', 'quick', 'quick', 'quick', 'quick', 'quick', 'quick',
    'expensive', 'cheap', 'expensive', 'cheap', 'expensive', 'cheap', 'expensive', 'cheap',
    'good', 'bad', 'good', 'bad', 'good', 'bad', 'good', 'bad', 'good', 'bad',
    'excellent', 'terrible', 'excellent', 'terrible', 'excellent', 'terrible', 'excellent',
    'very', 'very', 'very', 'very', 'very', 'very', 'very', 'very', 'very', 'very',
    'much', 'much', 'much', 'much', 'much', 'much', 'much', 'much', 'much', 'much',
    'little', 'little', 'little', 'little', 'little', 'little', 'little', 'little',
    'enough', 'enough', 'enough', 'enough', 'enough', 'enough', 'enough', 'enough',
    'too', 'too', 'too', 'too', 'too', 'too', 'too', 'too', 'too', 'too', 'too',
    'not', 'not', 'not', 'not', 'not', 'not', 'not', 'not', 'not', 'not', 'not',
    'nothing', 'nothing', 'nothing', 'nothing', 'nothing', 'nothing', 'nothing', 'nothing',
    'never', 'never', 'never', 'never', 'never', 'never', 'never', 'never', 'never',
    'always', 'always', 'always', 'always', 'always', 'always', 'always', 'always',
    'still', 'still', 'still', 'still', 'still', 'still', 'still', 'still', 'still',
    'already', 'already', 'already', 'already', 'already', 'already', 'already', 'already',
    'soon', 'soon', 'soon', 'soon', 'soon', 'soon', 'soon', 'soon', 'soon', 'soon',
    'now', 'now', 'now', 'now', 'now', 'now', 'now', 'now', 'now', 'now', 'now',
    'today', 'today', 'today', 'today', 'today', 'today', 'today', 'today', 'today',
    'yesterday', 'yesterday', 'yesterday', 'yesterday', 'yesterday', 'yesterday', 'yesterday',
    'tomorrow', 'tomorrow', 'tomorrow', 'tomorrow', 'tomorrow', 'tomorrow', 'tomorrow',
}

def detect_language(text: str) -> str:
    """
    Detect language from text.
    Returns: 'kinyarwanda', 'french', or 'english'
    """
    if not text or len(text.strip()) == 0:
        return 'english'  # default
    
    text_lower = text.lower().strip()
    words = re.findall(r'\b\w+\b', text_lower)
    
    if not words:
        return 'english'
    
    # Count matches for each language
    kinyarwanda_count = sum(1 for w in words if w in KINYARWANDA_WORDS)
    french_count = sum(1 for w in words if w in FRENCH_WORDS)
    english_count = sum(1 for w in words if w in ENGLISH_WORDS)
    
    # Calculate percentages
    total_matches = kinyarwanda_count + french_count + english_count
    
    if total_matches == 0:
        # No matches, use heuristics
        # Check for Kinyarwanda patterns (e.g., words ending in -a, -e, -i, -o, -u)
        if re.search(r'\b\w+[aeiou]$', text_lower):
            return 'kinyarwanda'
        return 'english'
    
    # Determine language by highest count
    if kinyarwanda_count > french_count and kinyarwanda_count > english_count:
        return 'kinyarwanda'
    elif french_count > english_count:
        return 'french'
    else:
        return 'english'

if __name__ == '__main__':
    # Test
    test_cases = [
        "mwaramutse",
        "hello",
        "bonjour",
        "how are you",
        "comment allez-vous",
        "waramutse gutese",
    ]
    
    for test in test_cases:
        lang = detect_language(test)
        print(f"'{test}' -> {lang}")
